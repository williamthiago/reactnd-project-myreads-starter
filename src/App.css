html, body, .root {
  height: 100%;
}
body {
  line-height: 1.5;
}
body, .app {
  background: white;
}

/* main page */

.list-books-title {
  position: fixed;
  width: 100%;
  z-index: 5;
  padding: 10px 0;
  background: #2e7c31;
  text-align: center;
}

.list-books-title h1 {
  font-weight: 400;
  margin: 0;
  color: white;
}

.list-books-content {
  margin-top: 65px;
  padding: 0 0 80px;
  flex: 1;
}

.bookshelf {
  padding: 0 10px 20px;
}

@media (min-width: 600px) {
  .bookshelf {
    padding: 0 20px 40px;
  }
}

.bookshelf-title {
  border-bottom: 1px solid #dedede;
}
.bookshelf-books {
  text-align: center;
}

.open-search {
  position: fixed;
  right: 25px;
  bottom: 25px;
}
.open-search a {
  display: block;
  width: 50px;
  height: 50px;
  border-radius: 50%;
  background: #2e7d32;
  background-image: url('./icons/add.svg');
  background-repeat: no-repeat;
  background-position: center;
  background-size: 28px;
  box-shadow: 0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23);
  font-size: 0;
}

/* search page */

.search-books-bar {
  position: fixed;
  width: 100%;
  top: 0;
  left: 0;
  z-index: 5;
  display: flex;
  box-shadow: 0 10px 20px rgba(0,0,0,0.19), 0 0 6px rgba(0,0,0,0.23);
}
.search-books-input-wrapper {
  flex: 1;
}
.search-books-bar input {
  width: 100%;
  padding: 15px 10px;
  font-size: 1.25em;
  border: none;
  outline: none;
}

.close-search {
  display: block;
  top: 20px;
  left: 15px;
  width: 50px;
  height: 53px;
  background: white;
  background-image: url('./icons/arrow-back.svg');
  background-position: center;
  background-repeat: no-repeat;
  background-size: 28px;
  font-size: 0;
}

.search-books-results {
  padding: 80px 10px 20px;
}

/* books grid */

.books-grid {
  list-style-type: none;
  padding: 0;
  margin: 0;

  display: flex;
  justify-content: center;
  flex-wrap: wrap;
}
.books-grid li {
  padding: 10px 15px;
  text-align: left;
}

.book {
  width: 140px;
}
.book-title,
.book-authors {
  font-size: 0.8em;
}
.book-title {
  margin-top: 10px;
}
.book-authors {
  color: #999;
}

.book-top {
  position: relative;
  height: 200px;
  display: flex;
  align-items: flex-end;
}

.book-shelf-changer {
  position: absolute;
  right: 0;
  bottom: -10px;
  width: 40px;
  height: 40px;
  border-radius: 50%;
  background: #60ac5d;
  background-image: url('./icons/arrow-drop-down.svg');
  background-repeat: no-repeat;
  background-position: center;
  background-size: 20px;
  box-shadow: 0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23);
}
.book-shelf-changer select {
  width: 100%;
  height: 100%;
  opacity: 0;
  cursor: pointer;
}

.book-tag {
  position: absolute;
  display: flex;  
  background: #999;
  align-items: center;
  padding: 0 5px 0 0;
  border-top-right-radius: 4px;
  border-bottom-right-radius: 4px;
  left: -4px;
  bottom: 7px;
  box-shadow: 1px 2px 0px rgba(0,0,0,.25);
  color: white;
  text-shadow: 1px 1px 0px rgba(0,0,0,0.1);
  cursor: pointer;
  z-index: 2;
}

.book-tag-label {
  padding: 3px 6px;
  font-size: 11px;
  display: block;
  border-right: 1px solid rgba(255,255,255,0.2);
  box-shadow: 1px 0px rgba(0,0,0,0.1);
}

.tag {
  display: table;
  margin-bottom: 8px;  
  padding: 3px 6px;
  border-radius: 4px;
  color: white;
  font-size: 11px;
  text-shadow: 1px 1px 0px rgba(0,0,0,0.1);
  cursor: pointer;
  z-index: 2;
}

.want-to-read {
  background: #ff7979;
}
.currently-reading {
  background: #29ace4;
}
.none {
  background: #999;
}
.read {
  background: #76af79;
}

.book-options {
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
  padding-right: 32px;
  display: flex;
  padding-bottom: 30px;
  align-items: flex-end;
  justify-content: flex-end;
  flex-direction: column;
  background: rgba(255,255,255,0.8);
  z-index: 1;
}

.book-options-toggle {
  display: block;
  transform: rotate(90deg);
  font-size: 24px;
  width: 0px;
  height: 0px;
  margin-top: -16px;
  margin-left: 35px;
  margin-right: -19px;
}

/* book cover */

.book-cover {
  box-shadow: 0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23);
  background: #eee;
}
.book-cover-title {
  padding: 20px 10px 0;
  text-align: center;
  font-size: 0.8em;
}


html, body, .app {
  min-height: 100%;
  min-height: 100vh;
}

.app {
  display: flex;
  flex-direction: column;
}

.loading-overlay {
  position: fixed;
  z-index: 10;

  top: 0;
  right: 0;
  width: 80px;
  height: 60px;
  overflow: hidden;
  display: flex;
  align-items: center;
  justify-content: center;
}

.loading-element {
  transform: scale(1.5);
}

/* Shimmer */

@-webkit-keyframes ShineAnimation{
  from {
      background-repeat:no-repeat;
      background-image:-webkit-linear-gradient(
          top left,
          rgba(255, 255, 255, 0.0) 0%,
          rgba(255, 255, 255, 0.0) 45%,
          rgba(255, 255, 255, 0.5) 48%,
          rgba(255, 255, 255, 0.8) 50%,
          rgba(255, 255, 255, 0.5) 52%,
          rgba(255, 255, 255, 0.0) 57%,
          rgba(255, 255, 255, 0.0) 100%
      );
      background-position:-250px -250px;
      background-size: 500px 600px
  }
  to {
      background-repeat:no-repeat;
      background-position:250px 250px;
  }
}

.shimmer {
  width:100%; /*Make sure the animation is over the whole element*/
  -webkit-animation-name: ShineAnimation;
  -webkit-animation-duration: 1.8s;
  -webkit-animation-iteration-count: infinite;
  -webkit-animation-timing-function: ease;

  background-color: rgb(225, 225, 225);
}

.shimmer-books {
  display:flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: center;
  align-items: flex-start;
}

.shimmer-book {
  margin: 20px 25px 0;
}

.shimmer-title {
  margin: 10px 20px;
  margin-top: 70px;
  padding-top: 20px;
  width: 250px;
  height: 30px;
}

.shimmer-cover {
  width: 130px;
  height: 180px;
}

.shimmer-text {
  max-height: 50%;
  width: 90px;
  height: 10px;
  margin: 5px 5px;
}
