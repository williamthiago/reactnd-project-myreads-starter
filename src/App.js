import React, { Component } from 'react'
import { Switch, Route } from 'react-router-dom'
import Snackbar from 'react-md/lib/Snackbars'
import ReactLoading from 'react-loading'

import './App.css'

import * as BooksAPI from './BooksAPI'
import Search from './components/Search'
import Library from './components/Library'

class App extends Component {
  state = {
    myBooks: [],
    loading: false,
    processing: false,
    toasts: []
  }

  async componentDidMount() {
    this.setState({ loading: true })
    const myBooks = await BooksAPI.getAll()
    this.setState({ myBooks, loading: false })
  }

  addToast = (text, action, autohide = true) => {
    this.setState((state) => {
      const toasts = state.toasts.slice()
      toasts.push({ text, action })
      return { toasts, autohide }
    });
  }

  dismissToast = () => {
    const [, ...toasts] = this.state.toasts
    this.setState({ toasts })
  };

  throwSyncError() {
    var newErr = new Error('Problems! Check your network, please')

    return err => { 
      newErr.originalError = err
      this.addToast(newErr.message)
      this.setState({
        processing: false
      })
      throw newErr
    }
  }

  async addBookToShelf(book, shelf) {
    this.setState({ processing: true })
    
    const currentShelvesState = await BooksAPI.update(book, shelf).catch(this.throwSyncError())

    const syncedShelves = Object.keys(currentShelvesState).reduce((bookIds, key) => {
      return currentShelvesState[key].reduce((bookIds, id) => {
        bookIds[id] = key
        return bookIds
      }, bookIds)
    }, {})

    const alreadyMine = !!this.state.myBooks.find(myBook => myBook.id === book.id)
    const filterMine = (book) => Object.keys(syncedShelves).includes(book.id)
    const syncShelf = (book) => ({
      ...book,
      shelf: syncedShelves[book.id]
    })

    let allBooks = alreadyMine ? this.state.myBooks : [...this.state.myBooks, book] 
    let myBooks = allBooks
      .filter(filterMine)
      .map(syncShelf)

    this.setState({
      myBooks,
      processing: false
    })

    this.addToast('Book updated!')
  }

  render() {
    const { myBooks, loading, toasts, processing } = this.state
    return (
      <div className="app">
        <Switch>
            <Route exact path="/" render={() => (
              <Library 
                books={myBooks} 
                loading={loading} 
                onChangeShelf={this.addBookToShelf.bind(this)} />
            )} />
            <Route path="/search" render={() => (
              <Search 
                processing={processing}
                books={myBooks} 
                onChangeShelf={this.addBookToShelf.bind(this)} />
            )} />
        </Switch>
        <Snackbar
          toasts={toasts}
          autohide={true}
          onDismiss={this.dismissToast}
        />
        {(processing || loading) && 
          <div className="loading-overlay">
            <ReactLoading className="loading-element" type="bubbles" color="#746994" delay={0} />
          </div>
        }
      </div>
    )
  }
}

export default App
