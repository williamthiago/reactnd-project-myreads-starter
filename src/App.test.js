import React from 'react'
import { shallow } from 'enzyme'

import App from './App'

it('Renders App without crashing', () => {
  expect(shallow(<App />)).toMatchSnapshot()
})