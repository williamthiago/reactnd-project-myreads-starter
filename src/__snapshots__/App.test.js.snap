// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders App without crashing 1`] = `
<div
  className="app"
>
  <Switch>
    <Route
      exact={true}
      path="/"
      render={[Function]}
    />
    <Route
      path="/search"
      render={[Function]}
    />
  </Switch>
  <SnackbarContainer
    autohide={true}
    autohideTimeout={3000}
    onDismiss={[Function]}
    toasts={Array []}
    transitionEnterTimeout={300}
    transitionLeaveTimeout={300}
    transitionName="md-snackbar"
  />
</div>
`;
